### PR TITLE
Cross-entropy loss without SoftmaxOutput

### PR DESCRIPTION
**Pull request by fhieber**
Use cross entropy loss implementation that avoids SoftmaxOutput by default.

Added tests to verify equivalence to SoftmaxOutput as much as possible (for the case without label smoothing). For label-smoothed cross-entropy, there actually seems to be an issue with SoftmaxOutputs gradients: I was not able to reproduce them with an independent implementation.

Speed-wise, the SoftmaxOutput-less cross-entropy loss is about 5-10% slower at equal memory cost.

However, given the issues with SoftmaxOutput smoothed gradients, I am inclined to change the default to a verifiable correct implementation.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by fhieber** [comment]
> Looks good!
> 
> Is the 5-10% slowdown for computing loss or for overall training throughput?

I had another look at the training logs. The slowdown is actually only 3-4% as measured in overall training throughput (time cost for until next checkpoint, or updates/sec)
